### PR TITLE
Fix build without HAVE_SERVER_PYTHON_PLUGINS

### PR DESCRIPTION
**Pull request by dmarteau**
## Description

Fix build without HAVE_SERVER_PYTHON_PLUGINS defined

This PR fix compile errors and link error when HAVE_SERVER_PYTHON_PLUGINS is undefined.

The main problem was  the fact that without HAVE_SERVER_PYTHON_PLUGINS then QgsAccessControl and QgsServerCacheManager are  forward declaration  only that do not resolve to any type.  Usually, this should not lead to problem as long as QgsAccessControl* should no be de-referenced nor any member accessed. Originally, this was an attempt to reduce the number of #ifdef/#endif directive in the code.

Subsequent refactoring has introduced new dependencies in QgsAccessControl and QgsServerCacheManager that access members at location not bracketed with #ifdef/#endif pair and thus leading to compile error.

This fix is not totally satisfactory in the sense that it reintroduce a bunch #define directive that does not contribute to readibility and maintenance.

Also this PR  fix the problem at hand, I have noticed that many patterns involving HAVE_SERVER_PYTHON_PLUGINS are the same in many places and that a cleaner refactoring of the 
code specific to server plugins should be possible.

### CODE REVIEW

File: src/server/services/wms/qgswmsgetlegendgraphics.cpp
Comment (m-kuhn) [inline src/server/services/wms/qgswmsgetlegendgraphics.cpp]: Unrelated suggestion

```suggestion
                                   QStringLiteral( "Output format '%1' is not supported in the GetLegendGraphic request" ).arg( format ) );
```

**Comment by m-kuhn** [comment]
Thanks a lot @dmarteau 

The code looks good to me to improve the current situation. A bigger refactoring is of course welcome, but I guess out of scope for now.

@mbernasocchi do you have the possibility to check if this branch works?

**Comment by rldhont** [comment]
Thanks @dmarteau 

Error during build.
```
src/server/qgsserverinterfaceimpl.cpp:32:16: error: member reference type 'QgsServerCacheManager *' is a pointer; did you mean to use '->'?
  mCacheManager.reset( new QgsServerCacheManager() );
```

Just replace 
```
mCacheManager.reset( new QgsServerCacheManager() );
```
By
```
mCacheManager = new QgsServerCacheManager();
```

File: src/server/qgsserverinterfaceimpl.cpp
Comment (elpaso) [inline src/server/qgsserverinterfaceimpl.cpp]: @dmarteau thanks! What was exactly the issue with the `unique_ptr` ?

Comment (dmarteau) [reply src/server/qgsserverinterfaceimpl.cpp]: @elpaso The unique_tr require that the destructor for the template parameter to be defined. In this case there is no destructor associated to the forward declaration and result in a link error.

**Review by rldhont — APPROVED** [review]

**Comment by mbernasocchi** [comment]
Thanks a lot @dmarteau
> @mbernasocchi do you have the possibility to check if this branch works?

on it now

**Comment by mbernasocchi** [comment]
@dmarteau, building with:
```
cmake /build/QGIS \
  -GNinja \
  -DWITH_STAGED_PLUGINS=OFF \
  -DCMAKE_INSTALL_PREFIX=/usr \
  -DWITH_GRASS=OFF \
  -DWITH_GRASS7=OFF \
  -DSUPPRESS_QT_WARNINGS=ON \
  -DENABLE_TESTS=OFF \
  -DWITH_QSPATIALITE=ON \
  -DWITH_QWTPOLAR=OFF \
  -DWITH_APIDOC=OFF \
  -DWITH_ASTYLE=OFF \
  -DWITH_DESKTOP=OFF \
  -DWITH_SERVER=ON \
  -DWITH_BINDINGS=OFF \
  -DWITH_QTMOBILITY=OFF \
  -DWITH_QUICK=OFF \
  -DWITH_3D=OFF \
  -DWITH_GUI=OFF \
  -DDISABLE_DEPRECATED=ON 

 ninja
```

I get

```
[1480/1550] Building CXX object src/server/CMakeFiles/qgis_server.dir/qgsserverplugins.cpp.o
FAILED: src/server/CMakeFiles/qgis_server.dir/qgsserverplugins.cpp.o 
ccache /usr/bin/c++  -DQGIS_DISABLE_DEPRECATED -DQT_CONCURRENT_LIB -DQT_CORE_LIB -DQT_DISABLE_DEPRECATED_BEFORE=0 -DQT_GUI_LIB -DQT_NETWORK_LIB -DQT_NO_CAST_TO_ASCII -DQT_NO_DEBUG -DQT_POSITIONING_LIB -DQT_PRINTSUPPORT_LIB -DQT_SERIALPORT_LIB -DQT_SQL_LIB -DQT_SVG_LIB -DQT_WEBKITWIDGETS_LIB -DQT_WEBKIT_LIB -DQT_WIDGETS_LIB -DQT_XML_LIB -DWITH_QTWEBKIT -Dqgis_server_EXPORTS -I. -isystem /usr/include/gdal -isystem /usr/include/postgresql -Isrc/server -isystem ../src/server/include/qgis -isystem /usr/include/Qca-qt5/QtCrypto -isystem /usr/include/qt5keychain -I../src/core -I../src/core/auth -I../src/core/dxf -I../src/core/expression -I../src/core/geometry -I../src/core/metadata -I../src/core/raster -I../src/core/symbology -I../src/core/layout -I../src/core/layertree -I../src/analysis/interpolation -I../src/plugins/diagram_overlay -I../src/python -Isrc/core -Isrc/python -Isrc/analysis -isystem /usr/include/x86_64-linux-gnu/qt5 -isystem /usr/include/x86_64-linux-gnu/qt5/QtXml -isystem /usr/include/x86_64-linux-gnu/qt5/QtCore -isystem /usr/lib/x86_64-linux-gnu/qt5/mkspecs/linux-g++ -isystem /usr/include/x86_64-linux-gnu/qt5/QtGui -isystem /usr/include/x86_64-linux-gnu/qt5/QtWidgets -isystem /usr/include/x86_64-linux-gnu/qt5/QtSvg -isystem /usr/include/x86_64-linux-gnu/qt5/QtPrintSupport -isystem /usr/include/x86_64-linux-gnu/qt5/QtNetwork -isystem /usr/include/x86_64-linux-gnu/qt5/QtWebKitWidgets -isystem /usr/include/x86_64-linux-gnu/qt5/QtWebKit -isystem /usr/include/x86_64-linux-gnu/qt5/QtSql -isystem /usr/include/x86_64-linux-gnu/qt5/QtConcurrent -isystem /usr/include/x86_64-linux-gnu/qt5/QtSerialPort -isystem /usr/include/x86_64-linux-gnu/qt5/QtPositioning -Wall -Wextra -Wno-long-long -Wformat-security -Wno-strict-aliasing -fPIC -fvisibility=hidden   -fPIC -std=gnu++11 -MD -MT src/server/CMakeFiles/qgis_server.dir/qgsserverplugins.cpp.o -MF src/server/CMakeFiles/qgis_server.dir/qgsserverplugins.cpp.o.d -o src/server/CMakeFiles/qgis_server.dir/qgsserverplugins.cpp.o -c ../src/server/qgsserverplugins.cpp
In file included from ../src/server/qgsserverplugins.cpp:22:0:
../src/python/qgspythonutils.h:22:10: fatal error: qgis_python.h: No such file or directory
 #include "qgis_python.h"
          ^~~~~~~~~~~~~~~
compilation terminated.
```

**Comment by dmarteau** [comment]
@mbernasocchi I Have tested that the build is ok with the options -DWITH_SERVER_PLUGINS=FALSE, nothing else. I do not not see that option in your build configuration.

**Comment by dmarteau** [comment]
@mbernasocchi Using -DWITH_GUI=OFF, I cannot event start the build on OSX:

`
ninja: error: '../python/plugins/db_manager/db_plugins/postgis/plugins/versioning/pygui', needed by 'python/plugins/db_manager/db_plugins/postgis/plugins/versioning/ui_DlgVersioning.py', missing and no known rule to make it
`

AMHA this is another matter than the WITH_SERVER_PLUGINS problem.

**Comment by dmarteau** [comment]
@mbernasocchi I confirm, that you need   -DWITH_SERVER_PLUGINS=FALSE, it is not trigged when WITH_BINDINGS is OFF. 

If  WITH_SERVER_PLUGINS is not set to FALSE,  you are trying to compile qgsserverplugins.cpp wich require python bindings.

File: CMakeLists.txt
Comment (m-kuhn) [inline CMakeLists.txt]: Just disabling server plugins when with bindings is off might also be confusing, in the end the problem is that we have two incompatible and conflicting options.

Maybe failing and giving a warning like here would be more appropriate?
https://github.com/qgis/QGIS/blob/700cfe8ab2641864185b09e08c0e2fad31155d8e/CMakeLists.txt#L144

Comment (dmarteau) [reply CMakeLists.txt]: Since WITH_SERVER_PLUGINS is ON by default, I would rather set a warning that server plugins are disabled.  
Because plugins support cannot be built without bindings (it is more a matter of hierarchy than conflicting options), I think would prefer that - as a logical dependency - instead requiring to disable plugins explicitely - but that's a matter of philosophy :-)

Comment (m-kuhn) [reply CMakeLists.txt]: Is the following compatible with both philosophies?

```
IF (WITH_SERVER)
  SET (WITH_SERVER_PLUGINS WITH_BINDINGS CACHE BOOL "Determines whether QGIS server support for python plugins should be built")
  IF(WITH_SERVER_PLUGINS AND NOT WITH_BINDINGS)
    MESSAGE(FATAL_ERROR "Smart message here")
  ENDIF(WITH_SERVER_PLUGINS AND NOT WITH_BINDINGS)
ENDIF (WITH_SERVER)
```

Comment (dmarteau) [reply CMakeLists.txt]: I'll go with it, but you still require to set both WITH_SERVER_PLUGINS=FALSE along  with WITH_BINDINGS=FALSE.

Comment (m-kuhn) [reply CMakeLists.txt]: I don't think you need to state both, `-DWITH_SERVER=FALSE -DWITH_BINDINGS=FALSE` will initialize `WITH_SERVER_PLUGINS` to false, no?

Comment (dmarteau) [reply CMakeLists.txt]: -DWITH_SERVER=FALSE disable SERVER_PLUGINS because you disable the whole server support. Furthemore there is no dependency between WITH_BINDINGS and WITH_SERVER.

You still need to state both WITH_SERVER_PLUGINS=FALSE and WITH_BINDINGS=FALSE but anyway its still better than a compile error.

Comment (m-kuhn) [reply CMakeLists.txt]: > You still need to state both WITH_SERVER_PLUGINS=FALSE and WITH_BINDINGS=FALSE but anyway its still better than a compile error.

I still cannot follow why. In the above example. `WITH_SERVER_PLUGINS` is initialized to the value of `WITH_BINDINGS`, if `WITH_BINDINGS` is false, `WITH_SERVER_PLUGINS` should implicitly also be false without any further ado. Or what am I missing?

Comment (dmarteau) [reply CMakeLists.txt]: My bad, I missed 'SET (WITH_SERVER_PLUGINS WITH_BINDINGS CACHE BOOL' !

Comment (dmarteau) [reply CMakeLists.txt]: mmh, Infortunately that doesn't seem to work: it seems that WITH_BINDINGS in the 'SET(WITH_SERVER_PLUGINS WITH_BINDINGS...) statement is not resolved as a variable.

Comment (dmarteau) [reply CMakeLists.txt]: The right syntax is:
`
SET (WITH_SERVER_PLUGINS ${WITH_BINDINGS} CACHE BOOL "..."
`

**Comment by rldhont** [comment]
@mbernasocchi or @m-kuhn can I merge it and backport it ?

**Comment by m-kuhn** [comment]
It looks good to me. @mbernasocchi could you test and build?

**Review by m-kuhn — APPROVED** [review]

**Comment by dmarteau** [comment]
I have rebased the branch with master, 
@mbernasocchi  @elpaso do you expect something else to do ?

**Comment by m-kuhn** [comment]
I would propose to merge this if there are no complaints by tomorrow and react if further issues are reported.

**Comment by m-kuhn** [comment]
@dmarteau I'm not sure my latest merge commit was really what was needed. Do you have an idea if that looks good?

**Comment by dmarteau** [comment]
May be I should have done a rebase before merging. I'm going to set up some functional tests to check the server state when compiled with  and without HAVE_SERVER_PYTHON_PLUGINS.

**Comment by m-kuhn** [comment]
@dmarteau do you prefer if

 * I undo my latest merge (force push to the state before)
 * I rebase your branch on master
 * Not do anything and leave it to you?

**Comment by dmarteau** [comment]
May be undo latest merge so I can rebase the branch and test compilation.

**Comment by m-kuhn** [comment]
done

**Comment by dmarteau** [comment]
Fixed conflicts

**Comment by m-kuhn** [comment]
Thanks a lot for your work @dmarteau !!

**Pull request merged**
